src/common/io.cc

```diff
@@ -120,27 +121,30 @@ std::string LoadSequentialFile(std::string fname) {
 #if defined(__linux__)
   posix_fadvise(fd, 0, 0, POSIX_FADV_SEQUENTIAL);
 #endif  // defined(__linux__)
-  ssize_t bytes_read = read(fd, &buffer[0], f_size_bytes);
-  if (bytes_read < 0) {
-    close(fd);
-    ReadErr();
+  ssize_t bytes_read = 0;
+  while (bytes_read < f_size_bytes) {
+    ssize_t result = read(fd, &buffer[bytes_read], f_size_bytes);
+    if (result < 0) {
+      close(fd);
+      ReadErr();
+    }
+    bytes_read += result;
   }
   close(fd);
-#else  // defined(__unix__)
-  FILE *f = fopen(fname.c_str(), "r");
-  if (f == NULL) {
-    std::string msg;
-    OpenErr();
+  buffer.back() = '\0';
+  // Trim NUL letters at the end
+  int64_t i;
+  for (i = static_cast<int64_t>(buffer.size()); i >= 0; --i) {
+    if (buffer[i - 1] != '\0') {
+      break;
+    }
   }
-  fseek(f, 0, SEEK_END);
-  auto fsize = ftell(f);
-  fseek(f, 0, SEEK_SET);
-
-  buffer.resize(fsize + 1);
-  fread(&buffer[0], 1, fsize, f);
-  fclose(f);
+  buffer = buffer.substr(0, i);
+#else  // defined(__unix__)
+  std::ifstream ifs(fname);
```

Comment:
Please note that this is terribly slow.

Reply:
Slower than `std::fread()`?

Reply:
In https://social.msdn.microsoft.com/Forums/en-US/87e128e2-5320-4796-b4a2-36248f86a8ce/huge-performance-hit-when-using-ifstream-vs-fopen?forum=Vsexpressvc, `ifstream` shows competitive performance with respect to `std::fread()`.

Reply:
> std::ifstream shows competitive performance with respect to std::fread().

Interesting as I just tried on Linux, it's about 10x slower.  But I guess when file is that huge JSON would be the bottleneck here.